services/nycrc/nycrc.service.js

```diff
@@ -0,0 +1,160 @@
+  }
+
+  static get route() {
+    return {
+      base: 'nycrc',
+      // TODO: eventually add support for .yml and .yaml:
+      pattern: ':user/:repo',
+      queryParamSchema: Joi.object({
+        config: Joi.string()
+          .regex(/(.*\.nycrc)|(.*\.json$)/)
+          .default('.nycrc'),
+        // Allow the default threshold detection logic to be overridden, .e.g.,
+        // favoring lines over branches:
+        preferredThreshold: Joi.string()
+          .optional()
+          .allow(...validThresholds),
+      }).required(),
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'nycrc config on GitHub',
+        namedParams: { user: 'yargs', repo: 'yargs' },
+        queryParams: { config: '.nycrc', preferredThreshold: 'lines' },
+        staticPreview: this.render({ coverage: 92 }),
+        documentation,
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'min coverage' }
+  }
+
+  static render({ coverage }) {
+    return {
+      message: `${coverage.toFixed(0)}%`,
+      color: coveragePercentage(coverage),
+    }
+  }
+
+  extractThreshold(config, preferredThreshold) {
+    const { branches, lines } = config
+    if (preferredThreshold) {
+      if (!validThresholds.includes(preferredThreshold)) {
+        throw new InvalidParameter({
+          prettyMessage: `threshold must be "branches", "lines", or "functions"`,
+        })
+      }
+      if (!config[preferredThreshold]) {
+        throw new InvalidResponse({
+          prettyMessage: `"${preferredThreshold}" threshold missing`,
+        })
+      }
+      return config[preferredThreshold]
+    } else if (branches || lines) {
+      // We favor branches over lines for the coverage badge, if both
+      // thresholds are provided (as branches is the stricter requirement):
+      return branches || lines
+    } else {
+      throw new InvalidResponse({
+        prettyMessage: '"branches" or "lines" threshold missing',
+      })
+    }
+  }
+
+  async handle({ user, repo }, queryParams) {
+    let coverage = NaN
+    const { config, preferredThreshold } = queryParams
+    if (config.includes('.nycrc')) {
+      coverage = this.extractThreshold(
+        await fetchJsonFromRepo(this, {
+          schema: nycrcSchema,
+          user,
+          repo,
+          branch: 'master',
+          filename: config,
+        }),
+        preferredThreshold
+      )
+    } else if (config.includes('package.json')) {
+      const pkgJson = await fetchJsonFromRepo(this, {
+        schema: pkgJSONSchema,
+        user,
+        repo,
+        branch: 'master',
+        filename: config,
+      })
+      const nycConfig = pkgJson.c8 || pkgJson.nyc
+      if (!nycConfig) {
+        throw new NotFound({
+          prettyMessage: 'no nyc or c8 stanza found',
+        })
+      } else {
+        coverage = this.extractThreshold(nycConfig, preferredThreshold)
+      }
+    }
+    return this.constructor.render({ coverage })
```

Comment:
Since an arbitrary file with a `.json` extension is accepted for the `config` query parameter, we should handle the case where someone provides something like `foo.json` (either by throwing an error or treating it like an nycrc config file and seeing if the key exists). 

As it stands currently, with such a value for `config` we'd display a `NaN%` message like ![](https://img.shields.io/badge/min%20coverage-NaN%25-brightgreen).

For example I deployed a [review app](https://shields-staging-pr-4759.herokuapp.com/) for this PR to test it out live, and was able to reproduce the NaN by using a repo that contains a shared nyc config module I use elsewhere:

https://shields-staging-pr-4759.herokuapp.com/nycrc/swellaby/nyc-config?config=partial-coverage.json
![](https://shields-staging-pr-4759.herokuapp.com/nycrc/swellaby/nyc-config?config=partial-coverage.json)

If it doesn't make sense for this badge to support any arbitrary json file (and I can understand why if so), then it'd be better IMO to throw an error instead of rendering NaN

Reply:
I think this scenario could also reasonably happen to end users when there's a typo in their config file name (`?config=.ncyrc.json` vs `?config=.nycrc.json`)

Reply:
@calebcartwright good catch. I opted to simply default to assuming `.json` files are in `.nycrc` format, unless we explicitly see `package.json`.